File: src/skillogy/infra/llm.py
"""LLM client provider with multi-provider auth.

Supported backends (auto-detected):
  1. Google Gemini   — when GOOGLE_API_KEY is set OR SKILLOGY_LLM=gemini
  2. Claude Agent SDK — when claude-agent-sdk is importable (inherits Claude Code auth)
  3. Anthropic API   — when ANTHROPIC_API_KEY is set

Override via SKILLOGY_LLM env var: "gemini" | "sdk" | "api".

Provider selection notes:
- Gemini path uses google-genai client.models.generate_content with system_instruction.
- Claude Agent SDK path uses the official query() async iterator. ClaudeAgentOptions
  does not expose a model selector — it uses whatever model Claude Code is configured
  with. The `model` argument is forwarded only on the API-key path and Gemini path.
"""

import importlib
import os
from typing import Optional

# Model ID normalization map
_MODEL_ALIASES = {
    "claude-haiku-4-5": "claude-haiku-4-5-20251001",
}

# Default Gemini model — `gemini-pro-latest` is an alias that always tracks the
# strongest Pro release (currently Gemini 3.1 Pro as of 2026-04-26). Override via env:
#   SKILLOGY_GEMINI_MODEL=gemini-flash-latest          (cheap/fast alias)
#   SKILLOGY_GEMINI_MODEL=gemini-flash-lite-latest     (cheapest alias)
#   SKILLOGY_GEMINI_MODEL=gemini-3.1-pro-preview       (explicit pin)
_DEFAULT_GEMINI_MODEL = os.environ.get("SKILLOGY_GEMINI_MODEL", "gemini-pro-latest")


def _normalize_model(model: str) -> str:
    return _MODEL_ALIASES.get(model, model)


def _is_sdk_available() -> bool:
    try:
        importlib.import_module("claude_agent_sdk")
        return True
    except ImportError:
        return False


def _is_genai_available() -> bool:
    try:
        importlib.import_module("google.genai")
        return True
    except ImportError:
        return False


class LLMClient:
    """Thin wrapper around an LLM backend providing a uniform complete() interface."""

    def complete(
        self,
        prompt: str,
        system: Optional[str] = None,
        max_tokens: int = 1024,
        temperature: float = 0.0,
    ) -> str:
        raise NotImplementedError


class _SDKClient(LLMClient):
    """Client backed by the official `claude-agent-sdk` query() iterator.

    Wraps the async `query()` API in a synchronous `complete()` call by running
    a fresh anyio event loop per call. For high-throughput indexing prefer the
    direct anthropic-SDK path (set SKILLOGY_FORCE_API_KEY=1).
    """

    def __init__(self, model: str) -> None:
        # Verify the SDK can be imported eagerly so failure surfaces at construction.
        # Actual imports done lazily inside complete() to keep test mocks simple.
        importlib.import_module("claude_agent_sdk")
        self._model = model

    def complete(
        self,
        prompt: str,
        system: Optional[str] = None,
        max_tokens: int = 1024,  # noqa: ARG002  (Agent SDK does not expose a max_tokens knob)
        temperature: float = 0.0,  # noqa: ARG002  (Agent SDK does not expose temperature)
    ) -> str:
        import anyio
        from claude_agent_sdk import (  # type: ignore[import-not-found]
            AssistantMessage,
            ClaudeAgentOptions,
            TextBlock,
            query,
        )

        options_kwargs: dict = {"max_turns": 1}
        if system is not None:
            options_kwargs["system_prompt"] = system
        options = ClaudeAgentOptions(**options_kwargs)

        async def _collect() -> str:
            chunks: list[str] = []
            async for message in query(prompt=prompt, options=options):
                if isinstance(message, AssistantMessage):
                    for block in message.content:
                        if isinstance(block, TextBlock):
                            chunks.append(block.text)
            return "".join(chunks)

        return anyio.run(_collect)


class _APIClient(LLMClient):
    """Client backed by the anthropic SDK using ANTHROPIC_API_KEY."""

    def __init__(self, model: str) -> None:
        from anthropic import Anthropic

        self._model = model
        self._client = Anthropic()

    def complete(
        self,
        prompt: str,
        system: Optional[str] = None,
        max_tokens: int = 1024,
        temperature: float = 0.0,
    ) -> str:
        kwargs: dict = {
            "model": self._model,
            "max_tokens": max_tokens,
            "temperature": temperature,
            "messages": [{"role": "user", "content": prompt}],
        }
        if system is not None:
            kwargs["system"] = system
        response = self._client.messages.create(**kwargs)
        return response.content[0].text


class _GeminiClient(LLMClient):
    """Client backed by google-genai (Gemini Developer API).

    Reads GOOGLE_API_KEY from the environment. Default model: gemini-2.5-flash.
    """

    def __init__(self, model: str) -> None:
        from google import genai  # type: ignore[import-not-found]

        self._model = model
        self._genai = genai
        # Client() reads GOOGLE_API_KEY automatically; allow explicit override.
        api_key = os.environ.get("GOOGLE_API_KEY")
        self._client = genai.Client(api_key=api_key) if api_key else genai.Client()

    def complete(
        self,
        prompt: str,
        system: Optional[str] = None,
        max_tokens: int = 1024,
        temperature: float = 0.0,
    ) -> str:
        from google.genai import types  # type: ignore[import-not-found]

        config_kwargs: dict = {
            "temperature": temperature,
            "max_output_tokens": max_tokens,
        }
        if system is not None:
            config_kwargs["system_instruction"] = system

        response = self._client.models.generate_content(
            model=self._model,
            contents=prompt,
            config=types.GenerateContentConfig(**config_kwargs),
        )
        return response.text or ""


def _resolve_model_for_provider(provider: str, model: str) -> str:
    """Translate caller's model id to the provider-appropriate one."""
    if provider == "gemini":
        return model if model.startswith("gemini") else _DEFAULT_GEMINI_MODEL
    return _normalize_model(model)


def get_llm_client(model: str = "claude-haiku-4-5") -> LLMClient:
    """Return an LLMClient using the best available auth method.

    Priority:
    1. SKILLOGY_LLM env var — explicit provider override ("gemini" | "sdk" | "api")
    2. GOOGLE_API_KEY set → Gemini (preferred for indexing throughput / free tier)
    3. claude-agent-sdk importable AND SKILLOGY_FORCE_API_KEY not set → SDK
    4. ANTHROPIC_API_KEY set → Anthropic API
    5. Raises RuntimeError with a clear message
    """
    forced = os.environ.get("SKILLOGY_LLM", "").lower()
    force_api_key = os.environ.get("SKILLOGY_FORCE_API_KEY")

    # 1. Explicit provider override
    if forced == "gemini":
        if not _is_genai_available():
            raise RuntimeError("SKILLOGY_LLM=gemini set but google-genai is not installed.")
        return _GeminiClient(model=_resolve_model_for_provider("gemini", model))
    if forced == "sdk":
        if not _is_sdk_available():
            raise RuntimeError("SKILLOGY_LLM=sdk set but claude-agent-sdk is not installed.")
        return _SDKClient(model=_normalize_model(model))
    if forced == "api":
        if not os.environ.get("ANTHROPIC_API_KEY"):
            raise RuntimeError("SKILLOGY_LLM=api set but ANTHROPIC_API_KEY is not set.")
        return _APIClient(model=_normalize_model(model))

    # 2. Auto-detect: Gemini first if key present (preferred for indexing)
    if os.environ.get("GOOGLE_API_KEY") and _is_genai_available():
        return _GeminiClient(model=_resolve_model_for_provider("gemini", model))

    # 3. Claude Agent SDK
    if _is_sdk_available() and not force_api_key:
        return _SDKClient(model=_normalize_model(model))

    # 4. Anthropic API
    if os.environ.get("ANTHROPIC_API_KEY"):
        return _APIClient(model=_normalize_model(model))

    raise RuntimeError(
        "No LLM auth available. Set GOOGLE_API_KEY (Gemini), install claude-agent-sdk"
        " (provided by Claude Code), or set ANTHROPIC_API_KEY."
    )

File: src/infra/llm.ts
/**
 * LLM client provider with multi-backend auth.
 *
 * Backends (auto-detected):
 *   1. Google Gemini   — when GOOGLE_API_KEY is set OR SKILLOGY_LLM=gemini
 *   2. Claude Agent SDK — inherits Claude Code OAuth (no API key needed)
 *   3. Anthropic API   — when ANTHROPIC_API_KEY is set
 *
 * Override via SKILLOGY_LLM env var: "gemini" | "sdk" | "api".
 */

const MODEL_ALIASES: Record<string, string> = {
  "claude-haiku-4-5": "claude-haiku-4-5-20251001",
};

const DEFAULT_GEMINI_MODEL =
  process.env.SKILLOGY_GEMINI_MODEL ?? "gemini-pro-latest";

function normalizeModel(model: string): string {
  return MODEL_ALIASES[model] ?? model;
}

export interface CompleteOptions {
  prompt: string;
  system?: string;
  maxTokens?: number;
  temperature?: number;
}

export interface LLMClient {
  complete(opts: CompleteOptions): Promise<string>;
}

class SDKClient implements LLMClient {
  // No model field — Claude Code's session controls which model the SDK uses.
  async complete(opts: CompleteOptions): Promise<string> {
    // Dynamic import keeps the dependency optional at module-load time.
    const { query } = (await import("@anthropic-ai/claude-agent-sdk")) as {
      query: (args: {
        prompt: string;
        options?: {
          systemPrompt?: string;
          maxTurns?: number;
          allowedTools?: string[];
        };
      }) => AsyncIterable<unknown>;
    };

    const queryOptions: { systemPrompt?: string; maxTurns: number; allowedTools: string[] } = {
      maxTurns: 1,
      allowedTools: [],
    };
    if (opts.system !== undefined) queryOptions.systemPrompt = opts.system;

    const chunks: string[] = [];
    for await (const message of query({ prompt: opts.prompt, options: queryOptions })) {
      const m = message as {
        type?: string;
        subtype?: string;
        result?: string;
        message?: { content?: Array<{ type?: string; text?: string }> };
      };

      if (m.type === "result" && m.subtype === "success" && typeof m.result === "string") {
        chunks.push(m.result);
        continue;
      }
      if (m.type === "assistant" && m.message?.content) {
        for (const block of m.message.content) {
          if (block.type === "text" && typeof block.text === "string") {
            chunks.push(block.text);
          }
        }
      }
    }
    return chunks.join("");
  }
}

class APIClient implements LLMClient {
  private _client: unknown;

  constructor(private readonly _model: string) {}

  async complete(opts: CompleteOptions): Promise<string> {
    if (!this._client) {
      const mod = (await import("@anthropic-ai/sdk")) as { default: new () => unknown };
      const Anthropic = mod.default;
      this._client = new Anthropic();
    }
    const client = this._client as {
      messages: {
        create: (args: {
          model: string;
          max_tokens: number;
          temperature: number;
          system?: string;
          messages: Array<{ role: "user"; content: string }>;
        }) => Promise<{ content: Array<{ type: string; text?: string }> }>;
      };
    };

    const args: Parameters<typeof client.messages.create>[0] = {
      model: this._model,
      max_tokens: opts.maxTokens ?? 1024,
      temperature: opts.temperature ?? 0.0,
      messages: [{ role: "user", content: opts.prompt }],
    };
    if (opts.system !== undefined) args.system = opts.system;

    const response = await client.messages.create(args);
    const first = response.content[0];
    if (first && first.type === "text" && typeof first.text === "string") {
      return first.text;
    }
    return "";
  }
}

class GeminiClient implements LLMClient {
  private _client: unknown;

  constructor(private readonly _model: string) {}

  async complete(opts: CompleteOptions): Promise<string> {
    if (!this._client) {
      const mod = (await import("@google/genai")) as {
        GoogleGenAI: new (args?: { apiKey?: string }) => unknown;
      };
      const GoogleGenAI = mod.GoogleGenAI;
      const apiKey = process.env.GOOGLE_API_KEY;
      this._client = apiKey ? new GoogleGenAI({ apiKey }) : new GoogleGenAI();
    }
    const client = this._client as {
      models: {
        generateContent: (args: {
          model: string;
          contents: string;
          config?: {
            temperature?: number;
            maxOutputTokens?: number;
            systemInstruction?: string;
          };
        }) => Promise<{ text?: string }>;
      };
    };

    const config: {
      temperature: number;
      maxOutputTokens: number;
      systemInstruction?: string;
    } = {
      temperature: opts.temperature ?? 0.0,
      maxOutputTokens: opts.maxTokens ?? 1024,
    };
    if (opts.system !== undefined) config.systemInstruction = opts.system;

    const response = await client.models.generateContent({
      model: this._model,
      contents: opts.prompt,
      config,
    });
    return response.text ?? "";
  }
}

interface ImportProbe {
  ok: boolean;
  // "missing" if the package itself isn't installed, "load-error" if it tried
  // to load but threw (most often: native binary mismatch on the SDK).
  reason?: "missing" | "load-error";
  message?: string;
}

async function probeImport(name: string): Promise<ImportProbe> {
  try {
    await import(name);
    return { ok: true };
  } catch (err) {
    const code = (err as NodeJS.ErrnoException).code;
    const message = (err as Error).message ?? String(err);
    if (code === "ERR_MODULE_NOT_FOUND" || /Cannot find package/.test(message)) {
      return { ok: false, reason: "missing", message };
    }
    return { ok: false, reason: "load-error", message };
  }
}

async function isImportable(name: string): Promise<boolean> {
  return (await probeImport(name)).ok;
}

function resolveModelForProvider(provider: string, model: string): string {
  if (provider === "gemini") {
    return model.startsWith("gemini") ? model : DEFAULT_GEMINI_MODEL;
  }
  return normalizeModel(model);
}

export async function getLlmClient(model = "claude-haiku-4-5"): Promise<LLMClient> {
  const forced = (process.env.SKILLOGY_LLM ?? "").toLowerCase();
  const forceApiKey = process.env.SKILLOGY_FORCE_API_KEY;

  // 1. Explicit override
  if (forced === "gemini") {
    if (!(await isImportable("@google/genai"))) {
      throw new Error("SKILLOGY_LLM=gemini set but @google/genai is not installed.");
    }
    return new GeminiClient(resolveModelForProvider("gemini", model));
  }
  if (forced === "sdk") {
    if (!(await isImportable("@anthropic-ai/claude-agent-sdk"))) {
      throw new Error(
        "SKILLOGY_LLM=sdk set but @anthropic-ai/claude-agent-sdk is not installed.",
      );
    }
    return new SDKClient();
  }
  if (forced === "api") {
    if (!process.env.ANTHROPIC_API_KEY) {
      throw new Error("SKILLOGY_LLM=api set but ANTHROPIC_API_KEY is not set.");
    }
    return new APIClient(normalizeModel(model));
  }

  // 2. Auto-detect — Gemini first when key + lib present
  if (process.env.GOOGLE_API_KEY && (await isImportable("@google/genai"))) {
    return new GeminiClient(resolveModelForProvider("gemini", model));
  }

  // 3. Claude Agent SDK — probe explicitly so we can surface load errors
  //    (e.g. missing platform-specific native binary) rather than silently
  //    falling through to "no auth".
  let sdkProbe: ImportProbe | undefined;
  if (!forceApiKey) {
    sdkProbe = await probeImport("@anthropic-ai/claude-agent-sdk");
    if (sdkProbe.ok) {
      return new SDKClient();
    }
    if (sdkProbe.reason === "load-error") {
      // The package is present but failed to load — usually the platform
      // native binary is missing or wrong arch. Print the cause; don't
      // silently fall back to "no auth".
      process.stderr.write(
        `[skillogy llm] @anthropic-ai/claude-agent-sdk failed to load: ${sdkProbe.message ?? "(no message)"}\n`,
      );
    }
  }

  // 4. Anthropic API
  if (process.env.ANTHROPIC_API_KEY) {
    return new APIClient(normalizeModel(model));
  }

  const sdkDetail =
    sdkProbe?.reason === "load-error"
      ? ` (claude-agent-sdk present but failed to load: ${sdkProbe.message ?? "?"})`
      : "";
  throw new Error(
    "No LLM auth available. Set GOOGLE_API_KEY (Gemini), install @anthropic-ai/claude-agent-sdk" +
      ` (provided by Claude Code), or set ANTHROPIC_API_KEY.${sdkDetail}`,
  );
}
